client/tokenized-payment-request/payment-request.js

```diff
@@ -0,0 +1,406 @@
+/* global jQuery, wcpayPaymentRequestParams */
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { doAction, addAction, removeAction } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import {
+	setPaymentRequestBranding,
+	trackPaymentRequestButtonClick,
+	trackPaymentRequestButtonLoad,
+} from './tracking';
+import {
+	transformStripePaymentMethodForStoreApi,
+	transformStripeShippingAddressForStoreApi,
+} from './transformers/stripe-to-wc';
+import {
+	transformCartDataForDisplayItems,
+	transformCartDataForShippingOptions,
+} from './transformers/wc-to-stripe';
+import paymentRequestButtonUi from './button-ui';
+import {
+	getPaymentRequest,
+	displayLoginConfirmationDialog,
+	getPaymentRequestData,
+	waitForAction,
+} from './frontend-utils';
+import PaymentRequestCartApi from './cart-api';
+import debounce from './debounce';
+
+const noop = () => null;
+
+/**
+ * Class to handle Stripe payment forms.
+ */
+export default class WcpayPaymentRequest {
+	/**
+	 * Whether the payment was aborted by the customer.
+	 */
+	paymentAborted = false;
+
+	/**
+	 * Whether global listeners have been added.
+	 */
+	areListenersInitialized = false;
+
+	/**
+	 * The cart data represented if the product were to be added to the cart (or, on cart/checkout pages, the cart data itself).
+	 * This is useful on product pages to understand if shipping is needed.
+	 */
+	cachedCartData = undefined;
+
+	/**
+	 * API to interface with the cart.
+	 *
+	 * @type {PaymentRequestCartApi}
+	 */
+	paymentRequestCartApi = undefined;
+
+	/**
+	 * WCPayAPI instance.
+	 *
+	 * @type {WCPayAPI}
+	 */
+	wcpayApi = undefined;
+
+	/**
+	 * On page load for product pages, we might get some data from the backend (which might get overwritten later).
+	 */
+	initialProductData = undefined;
+
+	constructor( { wcpayApi, paymentRequestCartApi, productData } ) {
+		this.wcpayApi = wcpayApi;
+		this.paymentRequestCartApi = paymentRequestCartApi;
+		this.initialProductData = productData;
+	}
+
+	/**
+	 * Starts the payment request
+	 */
+	async startPaymentRequest() {
+		// reference to this class' instance, to be used inside callbacks to avoid `this` misunderstandings.
+		const _self = this;
+		// TODO: is this creating multiple handlers to events on different `paymentRequest` objects?
```

Comment:
This is a note I added for myself to take a deeper investigation - the same logic is also happening in the current PRB implementation. We're inheriting the same code and logic, but it seems that it's been flying under the radar.

I did do a preliminary investigation on both implementations, and it does look like the memory heap increases & the iframe is cleared and re-mounted when the `init()` function is called in this scenario: https://github.com/Automattic/woocommerce-payments/blob/654097724c429f2ba812b94e7094c2c0f3b1df33/client/tokenized-payment-request/payment-request.js#L182-L182

Reply:
I forgot to mention - I had tried to call the `destroy()` method on the payment element, but it errored out.

Reply:
Is the fear here that we are making multiple requests to `stripe.paymentRequest()` via multiple summons of `WooPaymentsPaymentRequest::init()`? AFAICT, we should have a single instance of `WooPaymentsPaymentRequest`: is there anything we can add to this class to prevent this? 

I remember with the UPE, to prevent multiple mounts, we/I spammed multiple checks the payments elements object was defined before remounting. Can we add something similar to the `WooPaymentsPaymentRequest` class, if you feel this might be a significant enough concern?

Reply:
> Is the fear here that we are making multiple requests to stripe.paymentRequest() via multiple summons of WooPaymentsPaymentRequest::init()? 

Yes, that's part of the issue.
From the current PRB implementation, if the customer closes the PRB dialog a `paymentAborted` flag is set to `true`: https://github.com/Automattic/woocommerce-payments/blob/913339f61e410443dab710dad8506b5fb4983adb/client/payment-request/index.js#L268-L270

This flag is later checked when a variation attribute has changed, before attempting to update the PRB: https://github.com/Automattic/woocommerce-payments/blob/913339f61e410443dab710dad8506b5fb4983adb/client/payment-request/index.js#L476-L485
In that implementation, if the flow has been canceled, the PRB button is initialized again - making another request to `stripe.paymentRequest()`.

I am not sure why this is needed (in my testing I could just call [`paymentRequest.update()`](https://docs.stripe.com/js/payment_request/update) even after canceling the dialog without issue). My main concern was to keeping it to avoid possible issues/bugs 😅 🪄 
FWIW, I also didn't see anything in the Stripe documentation in regards to having to re-initialize the PRB element if the flow had cancelled.

What do you think - should we remove the `isPaymentAborted` check here ( https://github.com/Automattic/woocommerce-payments/blob/913339f61e410443dab710dad8506b5fb4983adb/client/tokenized-payment-request/payment-request.js#L171 ) to mitigate this TODO? 👀

Reply:
It sounds like we probably _could_ remove this check...but I don't really have enough knowledge to say for sure one way or the other. 😅 OK, it sounds to me like this is probably a valid TODO then and I'm content to leave this code as it is for now, but can we create a new issue to ensure that we investigate this separately? I just wanted to make sure that if we are leaving any TODOs in the code, they are actionable and won't just be forgotten. 👍